keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,733 @@
+IPv4, single-stack IPv6, or dual stack (with IPv4 primary). By
+contrast, `--node-ips ipv4` would mean single-stack IPv4 only (and
+would cause IPv6 addresses to be removed from `Node.Status.Addresses`)
+(Note that there is no way to say "dual-stack only".)
+
+```
+<<[UNRESOLVED optional-vs-required ]>>
+
+If we wanted to be able to have both "optionally dual-stack" and
+"mandatory dual-stack" we could have separate "optional ipv4/6" and
+"required ipv4/6" flags, but that seems to result in a lot of
+complication for not much benefit.
+
+<<[/UNRESOLVED]>>
+```
+
+If `--node-ips` is not specified but `--node-ip` is, then `--node-ips`
+is defaulted to an equivalent value as follows:
+
+  - If `--node-ip` is `0.0.0.0` then `--node-ips` is set to `ipv4`.
+
+  - If `--node-ip` is `::` then `--node-ips` is set to `ipv6`.
+
+  - Otherwise (when `--node-ip` is a specific IP address) `--node-ips`
+    is set to the same value as `--node-ip`.
+
+It is an error to pass both `--node-ip` and `--node-ips`.
+
+As discussed under [Implementation](#implementation),
+`Node.Status.Addresses` will then be set correctly to reflect this.
+Once `Node.Status.Addresses` has been set, the rule for actually
+finding the Primary Node IP is still the same as it used to be; we
+will just adjust `Node.Status.Addresses` to ensure that someone
+following the traditional rule will find the IP we want them to.
+
+There is now also a "Secondary Node IP", which is the first
+`InternalIP` address of the opposite IP family from the Primary Node
+IP, or the first such `ExternalIP` if there is no such `InternalIP`.
+
+#### Updated Use of `Node.Status.Addresses`
+
+When setting Pod status, kubelet will continue to use the Primary Node
+IP for `Pod.Status.HostIP`.
+
+In clusters with the `IPv6DualStack` feature gate enabled (or
+post-dual-stack GA), host-network Pods will now get both the Primary
+Node IP and the Secondary Node IP as `Pod.Status.PodIPs` (allowing
+host-network Pods to be endpoints of both IPv4 and IPv6 Services).
+
+Some methods in `utilnode` such as `GetNodeHostIP` and
+`GetPreferredNodeAddress` may be updated to have dual-stack versions
+that return multiple IPs. The existing versions are still useful on
+their own though.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+TODO
+
+## Design Details
+
+### Implementation
+
+As before, when using an external cloud provider, the external
+provider deals with setting `Node.Status.Addresses`, and when using a
+legacy cloud provider or bare metal, kubelet deals with setting it.
+
+#### With a Cloud Provider (External or Legacy)
+
+When using an external cloud provider, kubelet will pass the value of
```

Comment:
For the case of the kubelet with `--cloud-provider=external`, it actually falls back to look up the node IP (same behavior when no cloud provider is set) initially and then the external provider would override it afterwards. See https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/nodestatus/setters.go#L88-L93. 

Would/should that change any of the steps below?

Reply:
I was not planning to change that behavior; it would use the bare-metal code on first run but then defer to the cloud provider once the cloud provider set its own addresses